pkg/kubelet/kubelet.go

```diff
@@ -3070,3 +3051,83 @@ func (kl *Kubelet) PrepareDynamicResources(ctx context.Context, pod *v1.Pod) err
 func (kl *Kubelet) UnprepareDynamicResources(ctx context.Context, pod *v1.Pod) error {
 	return kl.containerManager.UnprepareDynamicResources(ctx, pod)
 }
+
+// staticPodRegistration ensures that all static pods are registered to the apiserver
+// before node became ready.
+func (kl *Kubelet) initialStaticPodsRegistration() error {
+	// kubelet running in standalone mode does not register static pods to the apiserver.
+	if kl.kubeClient == nil {
+		return nil
+	}
+	// Check if initial static pod registration has already been completed.
+	if kl.initialStaticPodsRegistered.Load() {
+		return nil
+	}
+	staticPodToMirrorPodMap := kl.podManager.GetStaticPodToMirrorPodMap()
+	for _, mirrorPod := range staticPodToMirrorPodMap {
+		if mirrorPod == nil {
+			return fmt.Errorf("not all static pods are registered")
+		}
+	}
+
+	kl.initialStaticPodsRegistered.Store(true)
+	return nil
+}
+
+// Ensure Mirror Pod for Static Pod exists.
+func (kl *Kubelet) ensureMirrorPodExists(staticPod, mirrorPod *v1.Pod) {
+	if !kubetypes.IsStaticPod(staticPod) {
+		return
+	}
+	deleted := false
+	if mirrorPod != nil {
+		if mirrorPod.DeletionTimestamp != nil || !kubepod.IsMirrorPodOf(mirrorPod, staticPod) {
+			// The mirror pod is semantically different from the static pod. Remove
+			// it. The mirror pod will get recreated later.
+			klog.InfoS("Trying to delete pod", "pod", klog.KObj(staticPod), "podUID", mirrorPod.ObjectMeta.UID)
+			podFullName := kubecontainer.GetPodFullName(staticPod)
+			var err error
+			deleted, err = kl.mirrorPodClient.DeleteMirrorPod(podFullName, &mirrorPod.ObjectMeta.UID)
+			if deleted {
+				klog.InfoS("Deleted mirror pod because it is outdated", "pod", klog.KObj(mirrorPod))
+			} else if err != nil {
+				klog.ErrorS(err, "Failed deleting mirror pod", "pod", klog.KObj(mirrorPod))
+			}
+		}
+	}
+	if mirrorPod == nil || deleted {
+		node, err := kl.GetNode()
+		if err != nil {
+			klog.V(4).ErrorS(err, "No need to create a mirror pod, since failed to get node info from the cluster", "node", klog.KRef("", string(kl.nodeName)))
+		} else if node.DeletionTimestamp != nil {
+			klog.V(4).InfoS("No need to create a mirror pod, since node has been removed from the cluster", "node", klog.KRef("", string(kl.nodeName)))
+		} else {
+			klog.V(4).InfoS("Creating a mirror pod for static pod", "pod", klog.KObj(staticPod))
+			if err := kl.mirrorPodClient.CreateMirrorPod(staticPod); err != nil {
```

Comment:
my understanding is that this will create the pod with that API server. How fast will we get it back in the `mirrorPodByFullName`? Should we expedite resyncing?

Reply:
Ideally we should _attempt_ to register and then mark the node as ready. I don't think we should keep waiting for mirror to be ACTUALLY created. It just opens door for all sorts of edge cases we haven't thought of

Reply:
> > Ideally we should attempt to register and then mark the node as ready. I don't think we should keep waiting for mirror to be ACTUALLY created.
> 
> The race condition causing OutOfCPU error is because we don't wait for the mirror pod to be created before the node is ready.

My understanding is that the OutOfCPU happening because of two things - we create static pod too late, and then do not wait for it's creation. My understanding of this PR is that we are trying to fix both - create mirrors fast, and also wait for them to mark the Node as ready. I have no issues with the first, it is safe enough for me. But I am worried that the second can cause troubles if for some reasons we will enter the loop of not being able to create a mirror pod at all. Like some admission webhook misconfiguration.

So if we are doing the waiting for status I would say we need:

1. To have a max wait timeout.
2. Protection from somebody constantly recreating new static pods.
3. Ability to turn the logic completely off (perhaps a feature gate)

Reply:
> My understanding is that the OutOfCPU happening because of two things - we create static pod too late, and then do not wait for it's creation.

The issue happens because mirror pods are created too late after node has become ready. 

> My understanding of this PR is that we are trying to fix both - create mirrors fast, and also wait for them to mark the Node as ready.

Correct.

> But I am worried that the second can cause troubles if for some reasons we will enter the loop of not being able to create a mirror pod at all. Like some admission webhook misconfiguration.

If mirror pods are not created ,then it can cause more issues like more pods scheduled on the node than it can handle. We would still see OutOfCPU issue in that case?

> 1. To have a max wait timeout.

It feels weird that we consider it as a problem only for a timeout duration. What would this timeout duration be? We would still see this OutOfCPU issue after the timeout period, so we are not fixing the issue?

> 2. Protection from somebody constantly recreating new static pods.

I have added this already. In `initialStaticPodsRegistration`, we only wait until initial static pods are registered to API server. 

> 3. Ability to turn the logic completely off (perhaps a feature gate)

Since this is a bug fix, does it make sense to keep it behind a feature gate?

Reply:
My problem here is that while fixing this issue we may introduce edge cases that will break customers in worse way than OutOfCPU. Imagine somebody with the bad static pod that somehow will fail creation. The cluster is working fine today, but with this bug fix they will end up not registering node AT ALL. This is way worse than one OutOfCPU issue, that may not be an issue for them at all.

Reply:
This is why I offering some ways we should consider as a backup safety options

Reply:
> Since this is a bug fix, does it make sense to keep it behind a feature gate?

we use to rollout changes that may impact visible behavior behind a feature gate as deprecated, so if we find out a few releases later that there is a problem we have a escape hatch without having to backport and tell usesrs to upgrade to the latest minor

> If mirror pods are not created ,then it can cause more issues like more pods scheduled on the node than it can handle.

agree with @SergeyKanzhelev here, not creating mirror pods is not as a big problem as not marking the node ready,  the mirror pod is just a representation of the pod that is already running in the node and functional ... I don't think is a bad idea to have a guardrail with a timeout and fall back to the previous behavior

Reply:
To resolve this concern, I propose splitting this PR into 2 PRs:

1. This PR only contains logic to register static pods quickly once node is registered. I feel this change should be safe to merge before 1.32 deadline and help reduce OutOfCPU issue.

2. Second PR will contain the logic to block node readiness on static pod registration. We can address this post-1.32 with more protection for safer release om future.

WDYT?

Reply:
> 1. This PR only contains logic to register static pods quickly once node is registered. I feel this change should be safe to merge before 1.32 deadline and help reduce OutOfCPU issue.

Yes, as long as the fix is reasonably contained. Definitely a good idea to move the first attempt of mirrors registration to right after the node was registered.

> 2. Second PR will contain the logic to block node readiness on static pod registration. We can address this post-1.32 with more protection for safer release om future.

Maybe if you are saying "temporarily block node readiness", it is fine. I don't think we can merge anything that will block the node readiness for a long time with the dependency on mirror pod creation being successful.

Reply:
I like the idea and I think that targeting 1. for this release is a good compromise ...

Reply:
+1 to prioritizing fast mirror pod creation.